packages-ts/gauntlet-terra-contracts/src/lib/contracts.ts

```diff
@@ -28,109 +29,164 @@ export type TerraABI = {
   [TERRA_OPERATIONS.QUERY]: JSONSchemaType<any>
 }
 
-export type Contract = {
-  id: CONTRACT_LIST
+export abstract class Contract {
+  // Contract metadata, initialized in constructor
+  readonly id: CONTRACT_LIST
+  readonly defaultVersion: string
+  readonly dirName: string
+  readonly downloadUrl: string
+
+  // Only load bytecode & schema later if needed
   abi: TerraABI
   bytecode: string
-}
+  addresses: AccAddress[] = []
+  address: AccAddress // first address, for convenience
+
+  constructor(id, dirName, defaultVersion) {
+    this.id = id
+    this.defaultVersion = defaultVersion
+    this.dirName = dirName
+  }
+
+  addInstance(name: string) {
+    const address = process.env[name]
+    if (!address) return this // it's okay if we don't have all the contract addresses, for now
+
+    this.addresses.push(address as AccAddress)
+    this.address = this.addresses[0]
+
+    if (!AccAddress.validate(address))
+      throw new Error(`Read invalid contract address ${address} for ${this.id} contract from env`)
+
+    logger.debug(`Deployed instance of ${this.id}: ${name}=${address}`)
+    return this
+  }
 
-export type Contracts = Record<CONTRACT_LIST, Contract>
+  loadContractCode = async (version: string): Promise<void> => {
+    if (version === 'local') {
+      // Possible paths depending on how/where gauntlet is being executed
+      const possibleContractPaths = [
+        path.join(__dirname, '../../artifacts/bin'),
+        path.join(process.cwd(), './artifacts/bin'),
+        path.join(process.cwd(), './tests/e2e/common_artifacts'),
+        path.join(process.cwd(), './packages-ts/gauntlet-terra-contracts/artifacts/bin'),
+      ]
+
+      const codes = possibleContractPaths
+        .filter((contractPath) => existsSync(`${contractPath}/${this.id}.wasm`))
+        .map((contractPath) => {
+          const wasm = readFileSync(`${contractPath}/${this.id}.wasm`)
+          return wasm.toString('base64')
+        })
+      this.bytecode = codes[0]
+    } else {
+      const url = `${this.downloadUrl}${version}/${this.id}.wasm`
+      logger.loading(`Fetching ${url}...`)
+      const response = await fetch(url)
+      const body = await response.arrayBuffer()
+      if (body.length == 0) {
+        throw new Error(`Download ${this.id}.wasm failed`)
+      }
+      this.bytecode = Buffer.from(body).toString('base64')
+    }
+  }
+
+  getContract = async (version: string): Promise<Contract> => {
```

Comment:
This seems more like a `create` function now, that basically creates the object. After that, calling `contract.getContract()` or `contract` will return the same.
Also, there will be incongruences if we "initialize" with one version, and call `getContract` later on with a different version.
Consider converting this into static `create` function, that initializes the object and return the instance